tensornetwork/block_sparse/block_tensor.py

```diff
@@ -0,0 +1,238 @@
+      the fusion of `charges` is non-zero.
+  Returns:
+    lookup: An np.ndarray of positive numbers between `0` and
+      `len(unique_charges)`. The position of values `n` in `lookup` are positions
+       with charge values `unique_charges[n]`.
+    unique_charges: The unique charges of fusion of `charges`
+    label_to_unique: The  integer labels of the unique charges.
+  """
+
+  fused_charges = fuse_charges(charges, flows)
+  unique_charges, inverse = fused_charges.unique(return_inverse=True)
+  _, label_to_unique, _ = unique_charges.intersect(
+      target_charges, return_indices=True)
+
+  tmp = np.full(len(unique_charges), fill_value=-1, dtype=np.int16)
+  tmp[label_to_unique] = label_to_unique
+  lookup = tmp[inverse]
+  lookup = lookup[lookup >= 0]
+
+  return lookup, unique_charges, label_to_unique
+
+
+def _get_strides(dims: Union[List[int], np.ndarray]) -> np.ndarray:
+  """
+  compute strides of `dims`.
+  """
+  return np.flip(np.append(1, np.cumprod(np.flip(dims[1::]))))
+
+
+def _find_best_partition(dims: List[int]) -> int:
+  """
+  Find the most-levelled partition of `dims`.
+  A levelled partitioning is a partitioning such that
+  np.prod(dim[:partition]) and np.prod(dim[partition:])
+  are as close as possible.
+  Args:
+    dims: A list or np.ndarray of integers.
+  Returns:
+    int: The best partitioning.
+  """
+  if len(dims) == 1:
+    raise ValueError(
+        'expecting `dims` with a length of at least 2, got `len(dims ) =1`')
+  diffs = [
+      np.abs(np.prod(dims[0:n]) - np.prod(dims[n::]))
+      for n in range(1, len(dims))
+  ]
+  min_inds = np.nonzero(diffs == np.min(diffs))[0]
+  if len(min_inds) > 1:
+    right_dims = [np.prod(dims[min_ind + 1:]) for min_ind in min_inds]
+    min_ind = min_inds[np.argmax(right_dims)]
+  else:
+    min_ind = min_inds[0]
+  return min_ind + 1
+
+
+def compute_fused_charge_degeneracies(charges: List[BaseCharge],
+                                      flows: List[bool]
+                                     ) -> Tuple[BaseCharge, np.ndarray]:
+  """
+  For a list of charges, computes all possible fused charges resulting
+  from fusing `charges` and their respective degeneracies
+  Args:
+    charges: List of `BaseCharge`, one for each leg of a 
+      tensor. 
+    flows: A list of bool, one for each leg of a tensor.
+      with values `False` or `True` denoting inflowing and 
+      outflowing charge direction, respectively.
+  Returns:
+    BaseCharge: The unique fused charges.
+    np.ndarray: The degeneracies of each unqiue fused charge.
+  """
+  if len(charges) == 1:
+    return (charges[0] * flows[0]).unique(return_counts=True)
+
+  # get unique charges and their degeneracies on the first leg.
+  # We are fusing from "left" to "right".
+  accumulated_charges, accumulated_degeneracies = (charges[0] *
+                                                   flows[0]).unique(
+                                                       return_counts=True)
```

Comment:
nit:
```python
accumulated_charges, accumulated_degeneracies = (
  charges[0] * flows[0]).unique(return_counts=True)
```
is more readable.

Reply:
yeah, my yapf configuration in emacs somehow makes these terrible line breaks. Haven't figured out how to change this yet